File: bindings/python/rrc/node.py
# SPDX-License-Identifier: 0BSD

from __future__ import annotations

from . import ffi
from .errors import Error, map_code
from .identity import Identity


class Node:
    def __init__(self, handle: int) -> None:
        self._handle = int(handle)

    @classmethod
    def create(cls, config_path: str = "") -> Node:
        h = ffi.lib.rrc_node_create(config_path.encode("utf-8"))
        if h == 0:
            raise Error(Error.INTERNAL)
        return cls(h)

    def start(self) -> None:
        map_code(ffi.lib.rrc_node_start(self._handle))

    def stop(self) -> None:
        map_code(ffi.lib.rrc_node_stop(self._handle))

    def set_identity(self, identity: Identity) -> None:
        map_code(ffi.lib.rrc_node_set_identity(self._handle, identity.handle))

    def add_udp_interface(self, name: str, local_addr: str, peer_addr: str) -> None:
        map_code(
            ffi.lib.rrc_node_add_udp_interface(
                self._handle,
                name.encode("utf-8"),
                local_addr.encode("utf-8"),
                peer_addr.encode("utf-8"),
            )
        )

    def has_path(self, dest_hash: bytes) -> bool:
        if len(dest_hash) != ffi.HASH_LEN:
            raise Error(Error.INVALID_ARG)
        arr = (ffi.ctypes.c_uint8 * ffi.HASH_LEN).from_buffer_copy(dest_hash)
        out = ffi.ctypes.c_int(0)
        map_code(
            ffi.lib.rrc_node_has_path(
                self._handle,
                arr,
                ffi.HASH_LEN,
                ffi.ctypes.byref(out),
            )
        )
        return bool(out.value)

    @property
    def handle(self) -> int:
        return self._handle

    def close(self) -> None:
        if self._handle:
            ffi.lib.rrc_node_destroy(self._handle)
            self._handle = 0

    def __enter__(self) -> Node:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass

File: bindings/java/src/io/quad4/rrc/Node.java
// SPDX-License-Identifier: 0BSD

package io.quad4.rrc;

import com.sun.jna.ptr.IntByReference;
import io.quad4.rrc.ffi.RrcLibrary;

public final class Node implements AutoCloseable {
    private long handle;

    private Node(long handle) {
        this.handle = handle;
    }

    public static Node create() {
        return create("");
    }

    public static Node create(String configPath) {
        long h = RrcLibrary.INSTANCE.rrc_node_create(configPath);
        if (h == 0) {
            throw new RrcException(RrcException.INTERNAL);
        }
        return new Node(h);
    }

    public void start() {
        Rrc.check(RrcLibrary.INSTANCE.rrc_node_start(handle));
    }

    public void stop() {
        Rrc.check(RrcLibrary.INSTANCE.rrc_node_stop(handle));
    }

    public void setIdentity(Identity identity) {
        Rrc.check(RrcLibrary.INSTANCE.rrc_node_set_identity(handle, identity.handle()));
    }

    public void addUdpInterface(String name, String localAddr, String peerAddr) {
        Rrc.check(RrcLibrary.INSTANCE.rrc_node_add_udp_interface(handle, name, localAddr, peerAddr));
    }

    public boolean hasPath(byte[] destHash) {
        if (destHash.length != Rrc.HASH_LEN) {
            throw new RrcException(RrcException.INVALID_ARG);
        }
        IntByReference out = new IntByReference();
        Rrc.check(RrcLibrary.INSTANCE.rrc_node_has_path(handle, destHash, destHash.length, out));
        return out.getValue() != 0;
    }

    public long handle() {
        return handle;
    }

    @Override
    public void close() {
        if (handle != 0) {
            RrcLibrary.INSTANCE.rrc_node_destroy(handle);
            handle = 0;
        }
    }
}
